### DISCUSSION TITLE
Reference libraries from AspNetCore framework in PowerShell binary module

### DISCUSSION THREAD

**Discussion by yodasad** [PowerShell/PowerShell#23888] [Q&A] [answered]
I'm writing a PowerShell (v7) binary module for managing AspNetCore Data Protection keys, so my module references Microsoft.AspNetCore.DataProtection.dll from Microsoft ASP.NET Core framework. I'm getting the following error when importing the module:
```
Import-Module: Could not load file or assembly 'Microsoft.AspNetCore.DataProtection, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.
```
The only way to resolve this issue is to publish my module as a self-contained library, so that all dependencies are published alongside the module library (so it is the whole AspNetCore).

I tried to add `<GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>` in my module's `csproj` file to generate a runtimeconfig.json file with `Microsoft.AspNetCore.App` framework reference, but it seems that powershell does not use `deps.json` and `runtimeconfig.json` when loading module's binaries (see this [issue](https://github.com/PowerShell/PowerShell/issues/15232)).

I tried to modify pwsh's `runtimeconfig.json` file, but it contains `includedFrameworks` section and it is not possible to combine this section with `frameworks` section (maybe it is possible in framework dependent  disribution of pwsh).

Are there any other options to reference AspNetCore framework from a PowerShell binary module?

**Comment by rhubarb-geek-nz** [comment, marked as answer]
Been there, done that. See [rhubarb-geek-nz/AspNetForPowerShell](https://github.com/rhubarb-geek-nz/AspNetForPowerShell).

Welcome to dll hell. This is an overview of the [mappings](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/package.psd1) you need to maintain, it is a little of of date but you get the idea.

The version of PowerShell needs to match the version of DotNet and AspNet runtimes. Then multiply that by Windows, Linux glib/musl, MacOS,arm/intel, 32bit and 64bit and you have quite a few combinations to maintain.

Fortunately that is all done here, 

- 	[AspNetForPowerShell.csproj](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/AspNetForPowerShell.csproj) which then calls 
- 	[package.ps1](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/package.ps1) which in turn calls
- 	[package-win.ps1](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/package-win.ps1)
- 	[package-osx.sh](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/package-osx.sh)
- 	[package-linux.sh](https://github.com/rhubarb-geek-nz/AspNetForPowerShell/blob/main/AspNetForPowerShell/package-linux.sh)

With the goal of creating packages (msi, pkg, deb, rpm ) which install the AspNetCore runtime in %ProgramFiles%\PowerShell\7\Modules or /opt/microsoft/powershell/7/Modules as appropriate.

So don't put the AspNetCore runtime in the PSGallery, your package would need about 200 copies of it within your module to match which ever installation of PowerShell is running.

So if you must use PowerShell then consider the alternatives....

1. Use an installation where you share the AspNetCore runtime from %ProgramFiles%\dotnet\shared\Microsoft.AspNetCore.App, /usr/share/dotnet/shared/Microsoft.AspNetCore.App or wherever. This can be either by using Mariner or Alpine which share the runtimes in the host OS, or using PowerShell as a dotnet tool.
2. Write your program as a dotnet tool as a simple C# console app and use the PowerShell SDK to run scripts or similar within your app.
3. Run an app with the AspNetCore runtime within a docker container and communicate with it using PowerShell's Invoke-WebRequest or similar. 

I recommend starting by using the dotnet PowerShell tool as the execution environment, that should give you access to the installed AspNetCore runtime.
